### PR TITLE
fix: support vercel path truncation when branch name is long

### PR DESCRIPTION
**Pull request by alfetopito**
# Summary

Addresses https://github.com/cowprotocol/bff/pull/223#discussion_r3513512141

# Testing

Unit tests

## Summary by CodeRabbit

* **Bug Fixes**
  * Fixed authorization validation for Vercel preview hostnames, including correct handling of maximum-length cases.
  * Improved rules for distinguishing full versus truncated scope fragments for both branch and build preview hostnames.
* **Tests**
  * Added more test coverage for maximum-length Vercel branch and build preview hostnames, including truncated-scope scenarios.
* **Chores**
  * Bumped the app version to the next patch release.

### CODE REVIEW

File: apps/api/src/app/plugins/bffAuth.ts [hunk 1]
```diff
@@ -97,24 +97,52 @@ function isAuthorizedVercelHostname(
...
+    // Missing full suffix is only valid when Vercel hit the DNS label limit
+    (!hasFullSuffix && deployment.length !== maxDeploymentLabelLength)
+  ) {
     return false
   }
 
   if (deployment.startsWith(branchPrefix)) {
-    return deployment.length > branchPrefix.length + suffix.length
+    // Branch previews need branch text before the optional scope suffix
+    return deployment.length > branchPrefix.length + (hasFullSuffix ? suffix.length : 0)
```
[the remark this answers is not in the document]
Comment (alfetopito) [reply apps/api/src/app/plugins/bffAuth.ts]: Addressed on https://github.com/cowprotocol/bff/pull/224/commits/3eafa892233322650e0cdc5f62f6a72d40a05efe

**Review by fairlighteth — COMMENTED** [review]
<details>
<summary>⚠️ <strong>AI Review</strong> (Codex GPT-5, worked 11m): truncated branch previews still allow fake scope fragments</summary>

**Finding:** [BLOCKING] Truncated branch labels can still bypass the Vercel team-scope check

- Location: `apps/api/src/app/plugins/bffAuth.ts:131`
- Lifecycle: `verified-vulnerability`
- Severity: `P2 (Medium)`
- Harness used: `DeepSec PR mode + targeted code inspection`
- PR responsibility: `introduced/kept reachable by this PR`
- Attacker-controlled input: a Vercel project/branch hostname with the configured `branchProject` prefix.
- Security boundary: `/proxies` origin allowlist before credentialed proxy routes.
- Impact: a different Vercel account can craft a branch URL whose 63-character label ends with `-c` (or another prefix of `-cowswap-dev`) and pass the current check without the hostname containing a verifiable team/account scope.

This one is important: the new branch-preview rule treats any trailing prefix of `-${scope}` as proof that the Vercel scope survived truncation. But branch names are part of the same visible label, so `swap-dev-git-${"a".repeat(48)}-c.vercel.app` is indistinguishable from a branch name ending in `-c` with no scope fragment at all. The current helper returns `true` for that shape, and the added test at `apps/api/src/app/plugins/bffAuth.spec.ts:249` locks that behavior in as allowed.

Vercel documents the scope slug as the account/team slug in generated URLs, and its shortening docs say generated labels are trimmed from the end and team slugs may be removed when shortening is needed. A branch-controlled suffix fragment is therefore not a server-verifiable team signal.

**Suggested fix**
- Do not authorize truncated branch preview hostnames unless the full `-${scope}` suffix survives, or use a server-verifiable signal instead of parsing a truncated branch hostname.
- Keep/rework truncated build-preview support separately if needed; the build-id path is less forgeable because the build id is constrained to alphanumeric before the hyphenated scope fragment.
- Add a regression test that expects `403` for `https://swap-dev-git-${"a".repeat(48)}-c.vercel.app` when configured with `vercel:swap-dev:cowswap-dev:swap`.

<details>
<summary>Review scope and related context</summary>

- Rechecked the resolved CodeRabbit thread about missing scope fragments. The exact no-fragment case is fixed, but this branch-tail variant remains.
- DeepSec status: run completed. It produced one broader finding: `Origin header is used as authentication for credentialed proxy routes`. I’m not repeating that as the main PR blocker because the overall Origin-as-auth design already exists on `main`, but it reinforces why this allowlist bypass matters for `/proxies`.
- DeepSec breadcrumb: `run-deepsec-pr-mode.sh --pr 224 --github-repo cowprotocol/bff`, exit code `1`, artifact `/tmp/deepsec-comment-bff-22935.md`.
- References checked: [Vercel generated URL components](https://vercel.com/docs/deployments/generated-urls), [Vercel shortening behavior](https://vercel.com/kb/guide/why-is-my-vercel-deployment-url-being-shortened).

</details>

<details>
<summary>🤖 Prompt for AI agents</summary>

```text
Verify this finding against current PR head. Fix only if still valid, keep the patch minimal, and validate the targeted failure mode.

Context:
- apps/api/src/app/plugins/bffAuth.ts:131-135 accepts truncated branch previews when the 63-character deployment label ends with a prefix of "-cowswap-dev".
- A branch name can itself end with that prefix, e.g. "a".repeat(48) + "-c", producing swap-dev-git-<branch>.vercel.app with no verifiable team scope.
- Expected fix: do not allow truncated branch-preview hostnames unless the full "-scope" survives, or require another server-verifiable signal.
- Add/adjust tests so the fake branch-tail case returns 403.
```

</details>

_Generated using the `pr-review` and `security-review` skills from the [CoW Protocol skills repo](https://github.com/cowprotocol/skills)._

</details>

**Comment by alfetopito** [comment]
> ⚠️ **AI Review** (Codex GPT-5, worked 11m): truncated branch previews still allow fake scope fragments
> 
> **Finding:** [BLOCKING] Truncated branch labels can still bypass the Vercel team-scope check
> 
>     * Location: `apps/api/src/app/plugins/bffAuth.ts:131`
> 
>     * Lifecycle: `verified-vulnerability`
> 
>     * Severity: `P2 (Medium)`
> 
>     * Harness used: `DeepSec PR mode + targeted code inspection`
> 
>     * PR responsibility: `introduced/kept reachable by this PR`
> 
>     * Attacker-controlled input: a Vercel project/branch hostname with the configured `branchProject` prefix.
> 
>     * Security boundary: `/proxies` origin allowlist before credentialed proxy routes.
> 
>     * Impact: a different Vercel account can craft a branch URL whose 63-character label ends with `-c` (or another prefix of `-cowswap-dev`) and pass the current check without the hostname containing a verifiable team/account scope.
> 
> 
> This one is important: the new branch-preview rule treats any trailing prefix of `-${scope}` as proof that the Vercel scope survived truncation. But branch names are part of the same visible label, so `swap-dev-git-${"a".repeat(48)}-c.vercel.app` is indistinguishable from a branch name ending in `-c` with no scope fragment at all. The current helper returns `true` for that shape, and the added test at `apps/api/src/app/plugins/bffAuth.spec.ts:249` locks that behavior in as allowed.
> 
> Vercel documents the scope slug as the account/team slug in generated URLs, and its shortening docs say generated labels are trimmed from the end and team slugs may be removed when shortening is needed. A branch-controlled suffix fragment is therefore not a server-verifiable team signal.
> 
> **Suggested fix**
> 
>     * Do not authorize truncated branch preview hostnames unless the full `-${scope}` suffix survives, or use a server-verifiable signal instead of parsing a truncated branch hostname.
> 
>     * Keep/rework truncated build-preview support separately if needed; the build-id path is less forgeable because the build id is constrained to alphanumeric before the hyphenated scope fragment.
> 
>     * Add a regression test that expects `403` for `https://swap-dev-git-${"a".repeat(48)}-c.vercel.app` when configured with `vercel:swap-dev:cowswap-dev:swap`.
> 
> 
> Review scope and related context
> 
> 
> 🤖 Prompt for AI agents
> 
> ```
> Verify this finding against current PR head. Fix only if still valid, keep the patch minimal, and validate the targeted failure mode.
> 
> Context:
> - apps/api/src/app/plugins/bffAuth.ts:131-135 accepts truncated branch previews when the 63-character deployment label ends with a prefix of "-cowswap-dev".
> - A branch name can itself end with that prefix, e.g. "a".repeat(48) + "-c", producing swap-dev-git-<branch>.vercel.app with no verifiable team scope.
> - Expected fix: do not allow truncated branch-preview hostnames unless the full "-scope" survives, or require another server-verifiable signal.
> - Add/adjust tests so the fake branch-tail case returns 403.
> ```
> 
> _Generated using the `pr-review` and `security-review` skills from the [CoW Protocol skills repo](https://github.com/cowprotocol/skills)._

Addressed on https://github.com/cowprotocol/bff/pull/224/commits/a37588a5876da60e2acbacf85755e7a551e11372

**Review by fairlighteth — COMMENTED** [review]
<details>
<summary>✅ <strong>AI Review</strong> (Codex GPT-5, worked 8m): follow-up addressed</summary>

**Rechecked**
- Author response: [comment](https://github.com/cowprotocol/bff/pull/224#issuecomment-4891141598) pointing to `a37588a`.
- Code path: `apps/api/src/app/plugins/bffAuth.ts:isAuthorizedVercelHostname`.
- Test coverage: `apps/api/src/app/plugins/bffAuth.spec.ts` now expects `403` for `swap-dev-git-${"a".repeat(48)}-c.vercel.app`.
- Automated checks: `yarn test`, `yarn lint`, and `yarn build` pass on PR head `a37588a`.

**Result:** Fixed. The old branch-tail bypass no longer applies because branch preview hostnames now return true only when `hasFullSuffix` is true; the exact fake-scope case returns false, while full-scope branch previews and truncated build previews still pass.

<details>
<summary>Review scope and related context</summary>

- Targeted reproduction checked:
  - fake branch-tail host length: `63`
  - fake branch-tail authorized: `false`
  - full branch host authorized: `true`
  - truncated build host authorized: `true`
- DeepSec status: run completed on a detached PR-head worktree at `a37588a`. Run `20260707154508-aa65d446a076b253` produced one `MEDIUM` high-confidence `auth-bypass` finding: Origin header allowlist is used as authorization for secret-backed `/proxies` routes. This is the broader Origin-as-auth issue already present on `origin/main`, not a remaining failure of the branch-preview fix.
- DeepSec artifacts: `/tmp/deepsec-comment-bff-pr-224-deepsec-365167.md` and `/tmp/deepsec-bff-pr224-bffAuth-finding.json`.
- Not repeated as a fresh blocker here: the broader Origin-as-auth issue is pre-existing on `main`; this follow-up only rechecks whether the prior branch-preview finding remains valid.

</details>

<details>
<summary>🤖 Verification notes for AI agents</summary>

```text
Verify only the prior branch-preview finding against current code.

Prior finding:
- A 63-character branch label ending in a prefix of "-cowswap-dev" could pass as if the scope survived truncation.

Current verification:
- apps/api/src/app/plugins/bffAuth.ts now requires hasFullSuffix for branch previews.
- apps/api/src/app/plugins/bffAuth.spec.ts covers the fake "-c" branch-tail case as 403.
- Keep truncated build-preview behavior separate; current code still allows the covered build truncation case.
- DeepSec rerun completed and found only the broader pre-existing Origin-as-auth issue, not a remaining branch-preview bypass.
```

</details>

_Generated using the `pr-review` and `security-review` skills from the [CoW Protocol skills repo](https://github.com/cowprotocol/skills)._

</details>

**Review by fairlighteth — APPROVED** [review]
